okhttp/src/main/java/io/grpc/okhttp/OkHttpServerTransport.java

```diff
@@ -164,15 +165,17 @@ private void startIo(SerializingExecutor serializingExecutor) {
       // The socket implementation is lazily initialized, but had broken thread-safety 
       // for that laziness https://bugs.openjdk.org/browse/JDK-8278326. 
       // As a workaround, we lock to synchronize initialization with shutdown().
+      HandshakerSocketFactory.HandshakeResult result;
       synchronized (lock) {
         socket.setTcpNoDelay(true);
       }
-      HandshakerSocketFactory.HandshakeResult result =
-          config.handshakerSocketFactory.handshake(socket, Attributes.EMPTY);
+      synchronized (handShakeLock) {
```

Comment:
This does nothing. This method is only called once ever per `handShakeLock` instance, and the lock isn't used anywhere else.

Reply:
Write of size 4 at 0x00008dc081dc by thread T73 (mutexes: write M0):
    #0 sun.security.ssl.SSLSessionImpl.invalidate()V SSLSessionImpl.java:781 
    #1 sun.security.ssl.TransportContext.fatal(Lsun/security/ssl/Alert;Ljava/lang/String;ZLjava/lang/Throwable;)Ljavax/net/ssl/SSLException; TransportContext.java:386 
    #2 sun.security.ssl.TransportContext.fatal(Lsun/security/ssl/Alert;Ljava/lang/String;Ljava/lang/Throwable;)Ljavax/net/ssl/SSLException; TransportContext.java:303 
    #3 sun.security.ssl.TransportContext.fatal(Lsun/security/ssl/Alert;Ljava/lang/Throwable;)Ljavax/net/ssl/SSLException; TransportContext.java:298 
    #4 sun.security.ssl.SSLSocketImpl.handleException(Ljava/lang/Exception;)V SSLSocketImpl.java:1686 
    #5 sun.security.ssl.SSLSocketImpl.startHandshake(Z)V SSLSocketImpl.java:466 
    #6 sun.security.ssl.SSLSocketImpl.startHandshake()V SSLSocketImpl.java:427 
    #7 io.grpc.okhttp.OkHttpProtocolNegotiator.negotiate(Ljavax/net/ssl/SSLSocket;Ljava/lang/String;Ljava/util/List;)Ljava/lang/String; OkHttpProtocolNegotiator.java:100 
    #8 io.grpc.okhttp.TlsServerHandshakerSocketFactory.handshake(Ljava/net/Socket;Lio/grpc/Attributes;)Lio/grpc/okhttp/HandshakerSocketFactory$HandshakeResult; TlsServerHandshakerSocketFactory.java:58 
    #9 io.grpc.okhttp.OkHttpServerTransport.startIo(Lio/grpc/internal/SerializingExecutor;)V OkHttpServerTransport.java:171 
    #10 io.grpc.okhttp.OkHttpServerTransport.lambda$start$0(Lio/grpc/internal/SerializingExecutor;)V OkHttpServerTransport.java:159 
    #11 io.grpc.okhttp.OkHttpServerTransport$$Lambda$90.run()V ?? 

  Previous read of size 4 at 0x00008dc081dc by thread T76:
    #0 sun.security.ssl.SSLSessionImpl.setContext(Lsun/security/ssl/SSLSessionContextImpl;)V SSLSessionImpl.java:768 
    #1 sun.security.ssl.SSLSessionContextImpl.put(Lsun/security/ssl/SSLSessionImpl;)V SSLSessionContextImpl.java:196 
    #2 sun.security.ssl.Finished$T12FinishedConsumer.onConsumeFinished(Lsun/security/ssl/ServerHandshakeContext;Ljava/nio/ByteBuffer;)V Finished.java:606 
    #3 sun.security.ssl.Finished$T12FinishedConsumer.consume(Lsun/security/ssl/ConnectionContext;Ljava/nio/ByteBuffer;)V Finished.java:527 
    #4 sun.security.ssl.SSLHandshake.consume(Lsun/security/ssl/ConnectionContext;Ljava/nio/ByteBuffer;)V SSLHandshake.java:392 
    #5 sun.security.ssl.HandshakeContext.dispatch(BLjava/nio/ByteBuffer;)V HandshakeContext.java:443 
    #6 sun.security.ssl.HandshakeContext.dispatch(BLsun/security/ssl/Plaintext;)V HandshakeContext.java:421 
    #7 sun.security.ssl.TransportContext.dispatch(Lsun/security/ssl/Plaintext;)V TransportContext.java:183 
    #8 sun.security.ssl.SSLTransport.decode(Lsun/security/ssl/TransportContext;[Ljava/nio/ByteBuffer;II[Ljava/nio/ByteBuffer;II)Lsun/security/ssl/Plaintext; SSLTransport.java:172 
    #9 sun.security.ssl.SSLSocketImpl.decode(Ljava/nio/ByteBuffer;)Lsun/security/ssl/Plaintext; SSLSocketImpl.java:1511 
    #10 sun.security.ssl.SSLSocketImpl.readHandshakeRecord()I SSLSocketImpl.java:1421 
    #11 sun.security.ssl.SSLSocketImpl.startHandshake(Z)V SSLSocketImpl.java:456 
    #12 sun.security.ssl.SSLSocketImpl.startHandshake()V SSLSocketImpl.java:427 
    #13 io.grpc.okhttp.OkHttpProtocolNegotiator.negotiate(Ljavax/net/ssl/SSLSocket;Ljava/lang/String;Ljava/util/List;)Ljava/lang/String; OkHttpProtocolNegotiator.java:100 
    #14 io.grpc.okhttp.TlsServerHandshakerSocketFactory.handshake(Ljava/net/Socket;Lio/grpc/Attributes;)Lio/grpc/okhttp/HandshakerSocketFactory$HandshakeResult; TlsServerHandshakerSocketFactory.java:58 
    #15 io.grpc.okhttp.OkHttpServerTransport.startIo(Lio/grpc/internal/SerializingExecutor;)V OkHttpServerTransport.java:171 
    #16 io.grpc.okhttp.OkHttpServerTransport.lambda$start$0(Lio/grpc/internal/SerializingExecutor;)V OkHttpServerTransport.java:159 
    #17 io.grpc.okhttp.OkHttpServerTransport$$Lambda$90.run()V ??

Reply:
This looks like a JDK bug. I don't know what JDK was used when TSAN failed, but based on the line numbers it looks to be JDK 11 or earlier. And it adds the new session to the session cache before setting the context, which does seem racy.
https://github.com/openjdk/jdk/blob/da75f3c4ad5bdf25167a3ed80e51f567ab3dbd01/src/java.base/share/classes/sun/security/ssl/SSLSessionContextImpl.java#L175

JDK master does the same, without any extra synchronization. So I think we need to report this and get it fixed.